README.md
# Supported-Picker-Dialogs

Travis CI Badge
https://travis-ci.com/Ibotta/Supported-Picker-Dialogs.svg?token=a1Fz6AJc1ZbpXy1CH2gh&branch=master

<!-- Add CodeClimate Badge https://docs.codeclimate.com/docs/overview#badges -->
<!-- Add any relevant artifact version badges (like rubygems.org, npmjs.com, etc. -->

## Installation

## Usage

## Development


CHANGELOG.md
<!--
  To help members of the community, it's a good idea to maintain a
  changelog to help people understand changes over time in your
  project.
  
  Better yet, have CI help you maintain the changelog. Check out the
  atomic_cache project for an example.
  https://github.com/Ibotta/atomic_cache/blob/416680ba88916129a4f8eecb127ade458a9714b8/ci-helpers/gem_helper#L88-L103

  More information is available at http://keepachangelog.com/en/1.0.0/
-->


app/src/test/java/com/ibotta/android/support/pickerapp/ExampleUnitTest.kt
package com.ibotta.android.support.pickerapp

import org.junit.Test

import org.junit.Assert.*

/**
 * Example local unit test, which will execute on the development machine (host).
 *
 * See [testing documentation](http://d.android.com/tools/testing).
 */
class ExampleUnitTest {
    @Test
    fun addition_isCorrect() {
        assertEquals(4, 2 + 2)
    }
}


pickerdialogs/src/test/java/com/ibotta/android/support/pickerdialogs/ExampleUnitTest.java
package com.ibotta.android.support.pickerdialogs;

import org.junit.Test;

import static org.junit.Assert.*;

/**
 * Example local unit test, which will execute on the development machine (host).
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */
public class ExampleUnitTest {
    @Test
    public void addition_isCorrect() {
        assertEquals(4, 2 + 2);
    }
}

app/src/main/java/com/ibotta/android/support/pickerapp/MainActivity.kt
package com.ibotta.android.support.pickerapp

import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.widget.DatePicker
import android.widget.TimePicker
import com.ibotta.android.support.pickerdialogs.SupportedDatePickerDialog
import com.ibotta.android.support.pickerdialogs.SupportedTimePickerDialog
import kotlinx.android.synthetic.main.activity_main.*
import java.text.SimpleDateFormat
import java.util.*

/**
 * The purpose of this class is to show off the SupportedDatePickerDialog and SupportedTimePickerDialog in various modes.
 *
 * Created by Lucas Newcomer 12/19/18
 */
class MainActivity : AppCompatActivity(), SupportedDatePickerDialog.OnDateSetListener,
        SupportedTimePickerDialog.OnTimeSetListener {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        loadState(savedInstanceState)
        setContentView(R.layout.activity_main)

        setDate()
        setTime()

        spinnerDatePickerBtn.setOnClickListener { showSpinnerDatePickerDialog() }
        calendarDatePickerBtn.setOnClickListener { showCalendarDatePickerDialog() }
        spinnerTimePickerBtn.setOnClickListener { showSpinnerTimePickerDialog() }
        calendarTimePickerBtn.setOnClickListener { showClockTimePickerDialog() }
    }

    private fun loadState(savedInstanceState: Bundle?) {
        savedInstanceState?.let {
            currentDate = savedInstanceState.getSerializable(CURRENT_DATE_KEY) as Calendar
        }
    }

    private fun showSpinnerDatePickerDialog() {
        val year = currentDate.get(Calendar.YEAR)
        val month = currentDate.get(Calendar.MONTH)
        val dayOfMonth = currentDate.get(Calendar.DAY_OF_MONTH)
        SupportedDatePickerDialog(this, R.style.SpinnerDatePickerDialogTheme, this, year, month, dayOfMonth).show()
    }

    private fun showCalendarDatePickerDialog() {
        // I added two more constructors to the SupportDatePickerDialog that is not part of the normal DatePickerDialog. This just allows
        // you to send an entire Calendar object. I always found it annoying that you only are able send the three date values
        // instead of the whole Calendar object.
        val supportedDatePickerDialog = SupportedDatePickerDialog(this, this, currentDate)
        if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.LOLLIPOP) {
            // For below API 21 the default date picker is a spinner and to update the date picker to be a calendar you
            // have to do it this way. Otherwise for above API 21 the default is the Material Design calendar date picker.
            supportedDatePickerDialog.datePicker.calendarViewShown = true
            supportedDatePickerDialog.datePicker.spinnersShown = false
        }
        supportedDatePickerDialog.show()
    }

    private fun showSpinnerTimePickerDialog() {
        val month = currentDate.get(Calendar.MONTH)
        val dayOfMonth = currentDate.get(Calendar.DAY_OF_MONTH)
        SupportedTimePickerDialog(
            context = this,
            themeResId =  R.style.SpinnerTimePickerDialogTheme,
            timeSetListener = this,
            mInitialHourOfDay = month,
            mInitialMinute = dayOfMonth,
            mIs24HourView = true).show()
    }

    private fun showClockTimePickerDialog() {
        val hour = currentDate.get(Calendar.HOUR_OF_DAY)
        val minute = currentDate.get(Calendar.MINUTE)
        // For above API 21 the default implementation of a time picker is a clock. For below API 21 there is no way to
        // make the time picker a clock.
        SupportedTimePickerDialog(
            context = this,
            timeSetListener = this,
            mInitialHourOfDay = hour,
            mInitialMinute = minute,
            mIs24HourView = true).show()
    }


    override fun onDateSet(view: DatePicker, year: Int, month: Int, dayOfMonth: Int) {
        currentDate.set(Calendar.YEAR, year)
        currentDate.set(Calendar.MONTH, month)
        currentDate.set(Calendar.DAY_OF_MONTH, dayOfMonth)
        setDate()
    }

    override fun onTimeSet(view: TimePicker, hourOfDay: Int, minute: Int) {
        currentDate.set(Calendar.HOUR_OF_DAY, hourOfDay)
        currentDate.set(Calendar.MINUTE, minute)
        setTime()
    }

    private fun setDate() {
        val formatter = SimpleDateFormat(getString(R.string.date_format), Locale.getDefault())
        val date = getString(R.string.date_picker_row_title, formatter.format(currentDate.time))
        dateDisplay.text = date
    }

    private fun setTime() {
        val formatter = SimpleDateFormat(getString(R.string.time_format), Locale.getDefault())
        val time = getString(R.string.time_picker_row_title, formatter.format(currentDate.time))
        timeDisplay.text = time
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        outState?.putSerializable(CURRENT_DATE_KEY, currentDate)
    }

    companion object {
        private var currentDate = Calendar.getInstance()
        private const val CURRENT_DATE_KEY = "currentDate"
    }
}


pickerdialogs/src/main/java/com/ibotta/android/support/pickerdialogs/SupportedTimePickerDialog.kt
package com.ibotta.android.support.pickerdialogs

import android.annotation.TargetApi
import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.os.Build
import android.os.Bundle
import android.util.TypedValue
import android.view.LayoutInflater
import android.widget.TimePicker

/**
 * The purpose of this class is to support all types of time picker dialogs for all types of Android versions. The reason
 * that this class was created is because spinner time pickers are not working for Android 7.0 (API 24) ref: https://issuetracker.google.com/issues/37119315.
 * This class is almost identical to the TimePickerDialog, which is part of the Android Platform Library, except for some library specific
 * logic in the class and TimePickers's xml. For instance .
 *
 * Created by Lucas Newcomer 12/17/18
 */
class SupportedTimePickerDialog @JvmOverloads constructor(
    context: Context,
    themeResId: Int = 0,
    private val timeSetListener: OnTimeSetListener,
    mInitialHourOfDay: Int,
    mInitialMinute: Int,
    mIs24HourView: Boolean
) : AlertDialog(context, resolveDialogTheme(context, themeResId)),
    DialogInterface.OnClickListener,
    TimePicker.OnTimeChangedListener {

    /**
     * @return the time picker displayed in the dialog
     * @hide For testing only.
     */
    private val timePicker: TimePicker

    /**
     * The callback interface used to indicate the user is done filling in
     * the time (e.g. they clicked on the 'OK' button).
     */
    interface OnTimeSetListener {
        /**
         * Called when the user is done setting a new time and the dialog has
         * closed.
         *
         * @param view the view associated with this listener
         * @param hourOfDay the hour that was set
         * @param minute the minute that was set
         */
        fun onTimeSet(view: TimePicker, hourOfDay: Int, minute: Int)
    }

    init {
        val themeContext = getContext()
        val inflater = LayoutInflater.from(themeContext)

        timePicker = inflater.inflate(R.layout.view_time_picker_dialog, null) as TimePicker
        timePicker.setIs24HourView(mIs24HourView)
        timePicker.currentHour = mInitialHourOfDay
        timePicker.currentMinute = mInitialMinute
        timePicker.setOnTimeChangedListener(this)
        setView(timePicker)

        setButton(DialogInterface.BUTTON_POSITIVE, themeContext.getString(android.R.string.ok), this)
        setButton(DialogInterface.BUTTON_NEGATIVE, themeContext.getString(android.R.string.cancel), this)
    }

    override fun onTimeChanged(view: TimePicker, hourOfDay: Int, minute: Int) {
        /* do nothing */
    }

    override fun onClick(dialog: DialogInterface, which: Int) {
        when (which) {
            DialogInterface.BUTTON_POSITIVE -> {
                if (isInputValid()) {
                    // Clearing focus forces the dialog to commit any pending
                    // changes, e.g. typed text in a NumberPicker.
                    timePicker.clearFocus()
                    timeSetListener.onTimeSet(timePicker, timePicker.currentHour, timePicker.currentMinute)
                }
            }
            DialogInterface.BUTTON_NEGATIVE -> cancel()
        }
    }

    @TargetApi(Build.VERSION_CODES.O)
    private fun isInputValid() : Boolean = ((android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.O)
        || ((android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) && timePicker.validateInput()))

    /**
     * Sets the current time.
     *
     * @param hourOfDay The current hour within the day.
     * @param minuteOfHour The current minute within the hour.
     */
    fun updateTime(hourOfDay: Int, minuteOfHour: Int) {
        timePicker.currentHour = hourOfDay
        timePicker.currentMinute = minuteOfHour
    }

    override fun onSaveInstanceState(): Bundle {
        val state = super.onSaveInstanceState()
        state.putInt(HOUR, timePicker.currentHour)
        state.putInt(MINUTE, timePicker.currentMinute)
        state.putBoolean(IS_24_HOUR, timePicker.is24HourView)
        return state
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        val hour = savedInstanceState.getInt(HOUR)
        val minute = savedInstanceState.getInt(MINUTE)
        timePicker.setIs24HourView(savedInstanceState.getBoolean(IS_24_HOUR))
        timePicker.currentHour = hour
        timePicker.currentMinute = minute
    }

    companion object {
        private const val HOUR = "hour"
        private const val MINUTE = "minute"
        private const val IS_24_HOUR = "is24hour"

        @TargetApi(Build.VERSION_CODES.LOLLIPOP)
        private fun resolveDialogTheme(context: Context, resId: Int): Int = when ((resId == 0)) {
            true -> {
                val outValue = TypedValue()
                context.theme.resolveAttribute(android.R.attr.timePickerDialogTheme, outValue, true)
                outValue.resourceId
            }
            false -> resId
        }
    }
}


pickerdialogs/src/main/java/com/ibotta/android/support/pickerdialogs/SupportedDatePickerDialog.kt
package com.ibotta.android.support.pickerdialogs

import android.annotation.TargetApi
import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.os.Build
import android.os.Bundle
import android.support.annotation.StyleRes
import android.util.TypedValue
import android.view.LayoutInflater
import android.widget.DatePicker

import java.util.Calendar

/**
 * The purpose of this class is to support all types of date picker dialogs for all types of Android versions. The reason
 * that this class was created is because spinner date pickers are not working for Android 7.0 (API 24) ref: https://issuetracker.google.com/issues/37120178.
 * This class is almost identical to the DatePickerDialog, which is part of the Android Platform Library, except for some library specific
 * logic in the class and DatePicker's xml. For instance the attribute dialogMode="true" is not something that can be used on a normal DatePicker in xml.
 *
 * Created by Lucas Newcomer 12/17/18
 */
class SupportedDatePickerDialog private constructor(
    context: Context,
    @StyleRes themeResId: Int,
    private var dateSetListener: OnDateSetListener?,
    calendar: Calendar?,
    private var year: Int,
    private var monthOfYear: Int,
    private var dayOfMonth: Int
) : AlertDialog(context, resolveDialogTheme(context, themeResId)),
    DialogInterface.OnClickListener,
    DatePicker.OnDateChangedListener {

    /**
     * Returns the [DatePicker] contained in this dialog.
     *
     * @return the date picker
     */
    val datePicker: DatePicker

    /**
     * Creates a new date picker dialog for the current date using the parent
     * context's default date picker dialog theme.
     *
     * @param context the parent context
     */
    constructor(context: Context) : this(context, 0)

    /**
     * Creates a new date picker dialog for the current date.
     *
     * @param context        the parent context
     * @param themeResId     the resource ID of the theme against which to inflate
     * this dialog, or `0` to use the parent
     * `context`'s default alert dialog theme
     */
    constructor(context: Context, @StyleRes themeResId: Int)
        : this(context, themeResId, null, Calendar.getInstance(), -1, -1, -1)

    /**
     * Creates a new date picker dialog for the specified date using the parent
     * context's default date picker dialog theme.
     *
     * @param context    the parent context
     * @param listener   the listener to call when the user sets the date
     * @param calendar   the calendar object used to populate the DatePicker
     */
    constructor(
        context: Context,
        listener: OnDateSetListener,
        calendar: Calendar
    ) : this(context, 0, listener, calendar)

    /**
     * Creates a new date picker dialog for the specified date using the parent
     * context's default date picker dialog theme.
     *
     * @param context    the parent context
     * @param listener   the listener to call when the user sets the date
     * @param calendar   the calendar object used to populate the DatePicker
     */
    constructor(
        context: Context,
        @StyleRes themeResId: Int,
        listener: OnDateSetListener,
        calendar: Calendar
    ) : this(context, themeResId, listener, calendar, -1, -1, -1)


    /**
     * Creates a new date picker dialog for the specified date using the parent
     * context's default date picker dialog theme.
     *
     * @param context    the parent context
     * @param listener   the listener to call when the user sets the date
     * @param year       the initially selected year
     * @param month      the initially selected month (0-11 for compatibility with
     * [Calendar.MONTH])
     * @param dayOfMonth the initially selected day of month (1-31, depending
     * on month)
     */
    constructor(
        context: Context,
        listener: OnDateSetListener,
        year: Int,
        month: Int,
        dayOfMonth: Int
    ) : this(context, 0, listener, year, month, dayOfMonth)

    /**
     * Creates a new date picker dialog for the specified date.
     *
     * @param context     the parent context
     * @param themeResId  the resource ID of the theme against which to inflate
     * this dialog, or `0` to use the parent
     * `context`'s default alert dialog theme
     * @param listener    the listener to call when the user sets the date
     * @param year        the initially selected year
     * @param monthOfYear the initially selected month of the year (0-11 for
     * compatibility with [Calendar.MONTH])
     * @param dayOfMonth  the initially selected day of month (1-31, depending
     * on month)
     */
    constructor(
        context: Context,
        @StyleRes themeResId: Int,
        listener: OnDateSetListener,
        year: Int,
        monthOfYear: Int,
        dayOfMonth: Int
    ) : this(context, themeResId, listener, null, year, monthOfYear, dayOfMonth)

    init {
        val themeContext = getContext()
        val inflater = LayoutInflater.from(themeContext)

        datePicker = inflater.inflate(R.layout.view_date_picker, null) as DatePicker

        setView(datePicker)

        setButton(DialogInterface.BUTTON_POSITIVE, themeContext.getString(android.R.string.ok), this)
        setButton(DialogInterface.BUTTON_NEGATIVE, themeContext.getString(android.R.string.cancel), this)

        calendar?.let {
            year = it.get(Calendar.YEAR)
            monthOfYear = it.get(Calendar.MONTH)
            dayOfMonth = it.get(Calendar.DAY_OF_MONTH)
        }

        datePicker.init(year, monthOfYear, dayOfMonth, this)
    }

    override fun onDateChanged(view: DatePicker, year: Int, month: Int, dayOfMonth: Int) {
        datePicker.init(year, month, dayOfMonth, this)
    }

    /**
     * Sets the listener to call when the user sets the date.
     *
     * @param listener the listener to call when the user sets the date
     */
    fun setOnDateSetListener(listener: OnDateSetListener) {
        dateSetListener = listener
    }

    override fun onClick(dialog: DialogInterface, which: Int) {
        when (which) {
            DialogInterface.BUTTON_POSITIVE -> dateSetListener?.let {
                // Clearing focus forces the dialog to commit any pending
                // changes, e.g. typed text in a NumberPicker.
                datePicker.clearFocus()
                it.onDateSet(datePicker, datePicker.year, datePicker.month, datePicker.dayOfMonth)
            }
            DialogInterface.BUTTON_NEGATIVE -> cancel()
        }
    }

    /**
     * Sets the current date.
     *
     * @param year       the year
     * @param month      the month (0-11 for compatibility with
     * [Calendar.MONTH])
     * @param dayOfMonth the day of month (1-31, depending on month)
     */
    fun updateDate(year: Int, month: Int, dayOfMonth: Int) {
        datePicker.updateDate(year, month, dayOfMonth)
    }

    override fun onSaveInstanceState(): Bundle {
        val state = super.onSaveInstanceState()
        state.putInt(YEAR, datePicker.year)
        state.putInt(MONTH, datePicker.month)
        state.putInt(DAY, datePicker.dayOfMonth)
        return state
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        val year = savedInstanceState.getInt(YEAR)
        val month = savedInstanceState.getInt(MONTH)
        val day = savedInstanceState.getInt(DAY)
        datePicker.init(year, month, day, this)
    }

    /**
     * The listener used to indicate the user has finished selecting a date.
     */
    interface OnDateSetListener {
        /**
         * @param view the picker associated with the dialog
         * @param year the selected year
         * @param month the selected month (0-11 for compatibility with
         * [Calendar.MONTH])
         * @param dayOfMonth th selected day of the month (1-31, depending on
         * month)
         */
        fun onDateSet(view: DatePicker, year: Int, month: Int, dayOfMonth: Int)
    }

    companion object {
        private const val YEAR = "year"
        private const val MONTH = "month"
        private const val DAY = "day"

        @StyleRes @TargetApi(Build.VERSION_CODES.LOLLIPOP)
        private fun resolveDialogTheme(context: Context, @StyleRes themeResId: Int): Int = when ((themeResId == 0)) {
            true -> {
                val outValue = TypedValue()
                // android.R.attr.datePickerDialogTheme was added in API 21, so the outValue.resourceId will be 0, and the view will receive no dialog theme
                context.theme.resolveAttribute(android.R.attr.datePickerDialogTheme, outValue, true)
                outValue.resourceId
            }
            false -> themeResId
        }
    }
}
